src/test/liquidity_tests.cpp

```diff
@@ -0,0 +1,377 @@
+        });
+
+        // distribute 100 coins
+        CAmount totalRwd = 100*COIN;
+
+        int64_t nTimeBegin = GetTimeMicros();
+        CAmount distributed = cache.DistributeRewards(totalRwd,
+            [&cache] (CScript const & owner, DCT_ID tokenID) {
+                return cache.GetBalance(owner, tokenID);
+            },
+            [&cache] (CScript const & to, CTokenAmount amount) {
+                return cache.AddBalance(to, amount);
+            }
+        );
+        int64_t nTimeEnd = GetTimeMicros(); auto nTimeRwd = nTimeEnd - nTimeBegin;
+        printf("Rewarded %d pools with %d shares each: %.2fms \n", PoolCount, ProvidersCount, 0.001 * (nTimeRwd));
+
+        printf("Distributed: = %ld\n", distributed);
+        BOOST_CHECK(distributed == 9999000000); // always slightly less due to MINIMUM_LIQUIDITY & rounding
+
+        // check it
+        auto rwd25 = 25 * COIN / ProvidersCount;
+        auto rwd50 = 50 * COIN / ProvidersCount;
+        cache.ForEachPoolShare([&] (DCT_ID const & id, CScript const & owner) {
+//            if (id == RWD25/* || id == RWD50*/)
+//                printf("owner = %s: %s,\n", owner.GetHex().c_str(), cache.GetBalance(owner, DCT_ID{0}).ToString().c_str());
+
+            // check only first couple of pools and the last (zero)
+            if (id == RWD25 && owner != CScript(id.v * ProvidersCount)) { // first got slightly less due to MINIMUM_LIQUIDITY
+                CAmount rwd = cache.GetBalance(owner, DCT_ID{0}).nValue;
+                BOOST_CHECK(rwd == rwd25);
+            }
+            if (id == RWD50 && owner != CScript(id.v * ProvidersCount)) { // first got slightly less due to MINIMUM_LIQUIDITY
+                CAmount rwd = cache.GetBalance(owner, DCT_ID{0}).nValue;
+                BOOST_CHECK(rwd == rwd50);
+            }
+            if (id == DCT_ID{10} ) { // first got slightly less due to MINIMUM_LIQUIDITY
+                CAmount rwd = cache.GetBalance(owner, DCT_ID{0}).nValue;
+                BOOST_CHECK(rwd == 0);
+            }
+            return true;
+        });
+
+        // check trade comissions for one of pools
+        {
+            DCT_ID idPool = DCT_ID{1};
+            auto optPool = cache.GetPoolPair(idPool);
+            cache.ForEachPoolShare([&] (DCT_ID const & id, CScript const & owner) {
+                if (id != idPool)
+                    return false;
+
+                if (owner != CScript(id.v * ProvidersCount)) { // first got slightly less due to MINIMUM_LIQUIDITY
+                    CAmount rwdA = cache.GetBalance(owner, DCT_ID{optPool->idTokenA}).nValue;
+                    CAmount rwdB = cache.GetBalance(owner, DCT_ID{optPool->idTokenB}).nValue;
+                    BOOST_CHECK(rwdA == id.v * COIN / ProvidersCount);
+                    BOOST_CHECK(rwdB == id.v * COIN * 2 / ProvidersCount);
+                }
+                return true;
+            }, PoolShareKey{idPool, {}});
+        }
+    }
+}
+
+
+BOOST_AUTO_TEST_SUITE_END()
```

Comment:
This test only covered AddLiquidity feature, but haven't cover the RemoveLiquidity feature.

Reply:
This test is exactly for testing math. Math with boundary values (too large or too low) that was hard to simulate by functional python's rpc calls. Or, math of rewards (which is hard to set by rpc's too). 
"removepoolliquidity", from this point of view, is trivial and tested by "feature_poolpair_liquidity.py"